src/common/config.h

```diff
@@ -2,84 +2,161 @@
  * Copyright 2014-2019 by Contributors
  * \file config.h
  * \brief helper class to load in configures from file
- * \author Haoda Fu
+ * \author Haoda Fu, Hyunsu Cho
  */
 #ifndef XGBOOST_COMMON_CONFIG_H_
 #define XGBOOST_COMMON_CONFIG_H_
 
+#include <dmlc/logging.h>
 #include <cstdio>
-#include <cstring>
 #include <string>
 #include <istream>
 #include <fstream>
 #include <vector>
+#include <regex>
+#include <iterator>
 #include <utility>
 
 namespace xgboost {
 namespace common {
 /*!
  * \brief Implementation of config reader
  */
-class ConfigParse {
+class ConfigParser {
  public:
   /*!
-   * \brief constructor
-   * \param cfgFileName name of configure file
+   * \brief Constructor for INI-style configuration parser
+   * \param path path to configuration file
    */
-  explicit ConfigParse(const std::string &cfgFileName) {
-    fi_.open(cfgFileName);
-    if (fi_.fail()) {
-      LOG(FATAL) << "cannot open file " << cfgFileName;
-    }
+  explicit ConfigParser(const std::string& path)
+    : line_comment_regex_("^#"),
```

Comment:
what about lines that do not start with '#' but a sequence of spaces followed by '#' - would that be treated as comments as well? if not, should it be?

how about comments at the end of a config item - a la: `key = value # this is a value of key`

i'm not suggesting that we support new syntax - just making sure that we do not regress on the existing behavior, if it indeed is supported today.

Reply:
> what about lines that do not start with '#' but a sequence of spaces followed by '#' - would that be treated as comments as well?

Yes.

> how about comments at the end of a config item

Yes, it would be supported.

Reply:
See an example of a comment at the end of an item: https://github.com/dmlc/xgboost/pull/4521/files/f376752b4a128519ad12246f50d162eaca0c0b22#diff-0c20fe270f3630af98c7b4428807af8cR110